File: 652. Find Duplicate Subtrees.py
# 652. Find Duplicate Subtrees

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        ans = []
        dp = defaultdict(int)

        def dfs(root):
            if not root:
                return '#'
            
            val = f"{root.val},{dfs(root.left)},{dfs(root.right)}"
            dp[val] += 1
            if dp[val] == 2:
                ans.append(root)
            return val

        dfs(root)
        # print(dp)
        return ans








        # dp = defaultdict(list)
        # ans = []
        
        # def dfs(node):
        #     if not node:
        #         return "N"
            
        #     key = ','.join([str(node.val), dfs(node.left), dfs(node.right)])
        #     if len(dp[key]) == 1:
        #         ans.append(node)
        #     dp[key].append(node)
        #     return key
        
        # dfs(root)
        # return ans
            

File: 652. Find Duplicate Subtrees.cpp
// 652. Find Duplicate Subtrees

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        unordered_map<string, int> subTreeMap;
        vector<TreeNode*> ans;
        serialize(root, subTreeMap, ans);
        return ans;
    }

private:
    string serialize(TreeNode* node,  unordered_map<string, int>& subTreeMap, vector<TreeNode*>& ans) {
        if (!node) return "#";

        string val = to_string(node->val) + "," + serialize(node->left, subTreeMap, ans) + "," + serialize(node->right, subTreeMap, ans);
        subTreeMap[val]++;
        if (subTreeMap[val] == 2) {
            ans.push_back(node);
        }
        return val;
    }

};